src/wallet/wallet_tx_builder.h

```diff
@@ -5,24 +5,352 @@
 
 using namespace libzcash;
 
+int GetAnchorHeight(const CChain& chain, int anchorConfirmations);
+
 /**
- * A payment that has been resolved to send to a specific
- * recipient address in a single pool.
+ * A payment that has been resolved to send to a specific recipient address in a single pool. This
+ * is an internal type that represents both user-requested payment addresses and generated
+ * (internal) payments (like change).
  */
 class ResolvedPayment : public RecipientMapping {
 public:
     CAmount amount;
     std::optional<Memo> memo;
+    bool isInternal;
 
     ResolvedPayment(
             std::optional<libzcash::UnifiedAddress> ua,
             libzcash::RecipientAddress address,
             CAmount amount,
+            std::optional<Memo> memo,
+            bool isInternal) :
+        RecipientMapping(ua, address), amount(amount), memo(memo), isInternal(isInternal) {}
+};
+
+/**
+ * A requested payment that has not yet been resolved to a
+ * specific recipient address.
+ */
+class Payment {
+private:
+    PaymentAddress address;
+    CAmount amount;
+    std::optional<Memo> memo;
+public:
+    Payment(
+            PaymentAddress address,
+            CAmount amount,
             std::optional<Memo> memo) :
-        RecipientMapping(ua, address), amount(amount), memo(memo) {}
+        address(address), amount(amount), memo(memo) {}
+
+    const PaymentAddress& GetAddress() const {
+        return address;
+    }
+
+    CAmount GetAmount() const {
+        return amount;
+    }
+
+    const std::optional<Memo>& GetMemo() const {
+        return memo;
+    }
+};
+
+class Payments {
+private:
+    std::vector<ResolvedPayment> payments;
+    std::set<OutputPool> recipientPools;
+    CAmount t_outputs_total{0};
+    CAmount sapling_outputs_total{0};
+    CAmount orchard_outputs_total{0};
+public:
+    Payments(std::vector<ResolvedPayment> payments) {
+        for (const ResolvedPayment& payment : payments) {
+            AddPayment(payment);
+        }
+    }
+
+    void AddPayment(ResolvedPayment payment) {
+        std::visit(match {
+            [&](const CKeyID& addr) {
+                t_outputs_total += payment.amount;
+                recipientPools.insert(OutputPool::Transparent);
+            },
+            [&](const CScriptID& addr) {
+                t_outputs_total += payment.amount;
+                recipientPools.insert(OutputPool::Transparent);
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                sapling_outputs_total += payment.amount;
+                recipientPools.insert(OutputPool::Sapling);
+            },
+            [&](const libzcash::OrchardRawAddress& addr) {
+                orchard_outputs_total += payment.amount;
+                recipientPools.insert(OutputPool::Orchard);
+            }
+        }, payment.address);
+        payments.push_back(payment);
+    }
+
+    const std::set<OutputPool>& GetRecipientPools() const {
+        return recipientPools;
```

Comment:
Should this be returning a reference to a set that can change underneath the client code if `AddPayment` is called? (Declaring it as a `const` reference helps slightly by saying that the client shouldn't change it, but also obscures that it can change anyway.)

In Rust terms it's similar to returning a borrowed reference to a structure when a mutable reference still exists, which would not be allowed. Consider returning a copy, or documenting the issue.

Reply:
Yeah, that should be a copy. Thanks!